src/ripple/rpc/impl/Role.cpp

```diff
@@ -74,40 +79,61 @@ requestRole (Role const& required, Port const& port,
  * ADMIN and IDENTIFIED roles shall have unlimited resources.
  */
 bool
-isUnlimited (Role const& required, Port const& port,
-    Json::Value const&params, beast::IP::Endpoint const& remoteIp,
-    std::string const& user)
+isUnlimited (Role const& role)
 {
-    Role role = requestRole(required, port, params, remoteIp, user);
-
-    if (role == Role::ADMIN || role == Role::IDENTIFIED)
-        return true;
-    else
-        return false;
+    return role == Role::ADMIN || role == Role::IDENTIFIED;
 }
 
 bool
-isUnlimited (Role const& role)
+isUnlimited (Role const& required, Port const& port,
+    Json::Value const& params, beast::IP::Endpoint const& remoteIp,
+    std::string const& user)
 {
-    return role == Role::ADMIN || role == Role::IDENTIFIED;
+    return isUnlimited(requestRole(required, port, params, remoteIp, user));
 }
 
 Resource::Consumer
 requestInboundEndpoint (Resource::Manager& manager,
-    beast::IP::Endpoint const& remoteAddress,
-        Port const& port, std::string const& user)
+    beast::IP::Endpoint const& remoteAddress, Role const& role,
+    std::string const& user, std::string const& forwardedFor)
 {
-    if (isUnlimited (Role::GUEST, port, Json::Value(), remoteAddress, user))
-        return manager.newUnlimitedEndpoint (to_string (remoteAddress));
+    if (isUnlimited(role))
+        return manager.newUnlimitedEndpoint (remoteAddress);
 
-    return manager.newInboundEndpoint(remoteAddress);
+    return manager.newInboundEndpoint(remoteAddress, role == Role::PROXY,
+        forwardedFor);
 }
 
-bool
-isIdentified (Port const& port, beast::IP::Address const& remoteIp,
-        std::string const& user)
+std::string
+forwardedFor(http_request_type const& request)
 {
-    return ! user.empty() && ipAllowed (remoteIp, port.secure_gateway_ip);
+    static std::string const forStr{"for="};
+    static std::size_t const forStrLen{forStr.size()};
+
+    auto it = request.find("X-Forwarded-For");
+    if (it != request.end())
+    {
+        std::string forwardedFor = it->value().to_string();
+        std::size_t const pos = forwardedFor.find(',');
+        if (pos != forwardedFor.npos)
+            forwardedFor.resize(pos);
+        return forwardedFor;
+    }
+    it = request.find("Forwarded");
+    if (it != request.end())
+    {
+        std::stringstream ss(it->value().to_string());
```

Comment:
since we don't really care about all the parts of `Forwarded`, we could just do something like:
```
        static std::string const forStr{"for="};
        auto forStart = it->value().find(forStr);
        if (forStart != boost::beast::string_view::npos)
        {
            forStart += forStr.size();
            const auto forEnd = it->value().find(";", forStart);
            return it->value().substr(forStart, forEnd-forStart).to_string();
        }
```
..which I think avoids a couple of temp strings.

Reply:
is the field a list of semicolon-delimited values? There might be something in here that could help:
https://github.com/boostorg/beast/blob/develop/include/boost/beast/http/rfc7230.hpp#L34

Reply:
@vinniefalco I'd like to use beast::param_list, but the format of the "Forwarded" value uses ";" as a delimiter between field-value pairs, and not as a start-of-record as in rfc7230.

Reply:
@vinniefalco However, ext_list is suitable for simplyfing how I extract from X-Forwarded-For.